AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java

```diff
@@ -48,141 +49,191 @@
-        assertNull(message.getData());
-
-        assertNull(message.getImageFilePath());
-        assertNull(message.getImageUrl());
-
-        LeanplumInboxMessage invalidMessage = LeanplumInboxMessage.createFromJsonMap("messageId", map);
-        assertNull(invalidMessage);
-    }
+        assertEquals("5231495977893888##1", message1.getMessageId());
+        assertEquals("This is a test inbox message", message1.getTitle());
+        assertEquals("This is a subtitle", message1.getSubtitle());
+        assertNull(message1.getExpirationTimestamp());
+        assertFalse(message1.isRead());
+
+        assertEquals("4682943996362752##2", message2.getMessageId());
+        assertEquals("This is a second test message", message2.getTitle());
+        assertEquals("This is a second test message subtitle", message2.getSubtitle());
+        assertNull(message2.getExpirationTimestamp());
+        assertFalse(message2.isRead());
+      }
+    };
+
+    // Add callback to be able to validate.
+    Leanplum.getInbox().addChangedHandler(callback);
+
+    // Remove it afterwards so we don't get callbacks anymore.
+    Leanplum.getInbox().removeChangedHandler(callback);
+
+    // Validate message state.
+    List<LeanplumInboxMessage> messageList = Leanplum.getInbox().unreadMessages();
+
+    LeanplumInboxMessage message1 = messageList.get(0);
+    LeanplumInboxMessage message2 = messageList.get(1);
+
+    message1.read();
+    assertTrue(message1.isRead());
+    assertEquals(1, Leanplum.getInbox().unreadCount());
+
+    message2.read();
+    assertTrue(message2.isRead());
+    assertEquals(0, Leanplum.getInbox().unreadCount());
+
+    assertEquals(2, Leanplum.getInbox().count());
+
+    LeanplumInboxMessage messageById = Leanplum.getInbox().messageForId(message1.getMessageId());
+    assertNotNull(messageById);
+    assertEquals(message1, messageById);
+
+    Leanplum.getInbox().removeMessage(messageById.getMessageId());
+    assertEquals(1, Leanplum.getInbox().allMessages().size());
+
+    // Validate inbox synced callback with false when there is no internet .
+    InboxSyncedCallback inboxSyncedCallbackFalse = new InboxSyncedCallback() {
+      @Override
+      public void onForceContentUpdate(boolean success) {
+        assertFalse(success);
+      }
+    };
+
+    // Add synced callback  to be able to validate.
+    Leanplum.getInbox().addSyncedHandler(inboxSyncedCallbackFalse);
+
+    // Return false when checks for internet connection.
+    doReturn(false).when(Util.class, "isConnected");
+
+    // Download messages.
+    Leanplum.getInbox().downloadMessages();
+
+    //  Remove synced callback afterwards so we don't get synced callbacks anymore.
+    Leanplum.getInbox().removeSyncedHandler(inboxSyncedCallbackFalse);
+  }
+
+  @Test
+  public void testDisablePrefetching() {
+    leanplumInbox.disableImagePrefetching();
+    assertFalse(leanplumInbox.isInboxImagePrefetchingEnabled());
+  }
+
+  @Test
+  public void testInvalidMessageIdIsRejected() {
```

Comment:
Why is this one a separate test? Please put old lines 185-186 back and if you want, please put comment what those lines do.

Reply:
testInbox needs to be broken down into proper unit tests, it is a pseudo integration test. See my comment below on testing only a single behavior in each unit test.